WordPress/Classes/Utility/CoreDataManager.swift

```diff
@@ -0,0 +1,401 @@
+            self.sentryStartupError.add(error: error)
+
+            /// Backup the old Store
+            ///
+            do {
+                try CoreDataIterativeMigrator.backupDatabase(at: self.storeURL)
+            } catch {
+                self.sentryStartupError.add(error: error)
+                self.sentryStartupError.send(title: "Can't initialize Core Data stack")
+                fatalError("☠️ [CoreDataManager] Cannot backup Store: \(error)")
+            }
+
+            /// Retry!
+            ///
+            container.loadPersistentStores { [weak self] (storeDescription, error) in
+                guard let error = error as NSError? else {
+                    return
+                }
+
+                self?.sentryStartupError.add(error: error)
+                self?.sentryStartupError.send(title: "Can't initialize Core Data stack")
+                fatalError("☠️ [CoreDataManager] Recovery Failed! \(error) [\(error.userInfo)]")
+            }
+        }
+        return container
+    }()
+
+    private func childContext(with concurrencyType: NSManagedObjectContextConcurrencyType) -> NSManagedObjectContext {
+        let childManagedObjectContext = NSManagedObjectContext(concurrencyType: concurrencyType)
+        childManagedObjectContext.parent = self.mainContext
+        childManagedObjectContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        return childManagedObjectContext
+    }
+
+    func newDerivedContext() -> NSManagedObjectContext {
+        childContext(with: .privateQueueConcurrencyType)
+    }
+
+    func newMainContextChildContext() -> NSManagedObjectContext {
+        childContext(with: .mainQueueConcurrencyType)
+    }
+
+    func saveContextAndWait(_ context: NSManagedObjectContext) {
+        save(context, wait: true)
+    }
+
+    func save(_ context: NSManagedObjectContext) {
+        save(context, wait: false)
+    }
+
+    func save(_ context: NSManagedObjectContext, withCompletionBlock completionBlock: @escaping () -> Void) {
+        save(context, wait: false, completion: completionBlock)
+    }
+
+    private func save(_ context: NSManagedObjectContext, wait: Bool, completion: (() -> Void)? = nil) {
+
+        guard context.parent != self.mainContext else {
+            saveDerivedContext(context, wait: wait, completion: completion)
+            return
+        }
+
+        guard wait else {
+            context.perform {
+                self.performSave(context: context, completion: completion)
+            }
+            return
+        }
+        context.performAndWait {
+            self.performSave(context: context, completion: completion)
+        }
+    }
+
+    private func performSave(context: NSManagedObjectContext, completion: (() -> Void)? = nil) {
+
+        let objects = Array(context.insertedObjects)
+        do {
+            try context.obtainPermanentIDs(for: objects)
+
+            guard context.hasChanges else {
+                return
```

Comment:
In cases where the guard fails, or there is an error and the catch block executes, should we still all the completion block?  It seems like if we don't we could reach a dead end execution path in some parts of the app, similar to API service calls that return vs calling their success or failure methods.  Thoughts?

Reply:
Thanks for spotting this!! In fact it was actually happening..